Action Panel: Allow figure to be left aligned

pablinos opened the pull request:
As part of the project to overhaul the 'Earn' section, we need a card format that's similar to the action panel, but has the image on the lefthand side.

This change adds a prop to the figure component, that can be set to `left` or `right`, with a default of right to be consistent with the current behaviour.

#### Testing instructions

Using this branch, navigate to http://calypso.localhost:3000/devdocs/design/action-panel 

The example has been updated with two versions of the component, one with an image on the right, and the other using the new functionality to display it on the left.

#### N.B.

The figure is positioned using floats, which can lead to elements wrapping around the image if it's too small. This can't really be helped without updating the component's HTML. There should probably be a follow-up change to rework this after auditing the use of the `AuditPanel`

pablinos committed:
Action Panel: Allow figure to be left aligned

As part of the project to overhaul the 'Earn' section, we need a card
format that's similar to the action panel, but has the image on the
lefthand side.

This change adds a prop to the figure component, that can be set to
`left` or `right`, with a default of right to be consistent with the
current behaviour.

tyxla commented on client/components/action-panel/docs/example.jsx:
```diff
@@ -41,6 +41,30 @@ const ActionPanelExample = () => (
+						inviting the user to do.
+					</p>
+					<ActionPanelFooter>
+						<Button className="action-panel__support-button" href="/help/contact">
```
I know there are classname lint errors from before, but it would be lovely if we could fix them in this PR.

pablinos replied:
Yes, the lint errors seem to be because this file is called `example.jsx` but the class names don't conform to that component name. It would be great to fix that, I'll see if I can work out a tweak we can make to the rule.

tyxla replied:
Would be great, but don't let that block the PR. I'm fine with this one as-is, since we already had those lint errors there.

pablinos replied:
For now, I've disabled the rule in this file. Looking at the code for the rule, I think we could change it so it does something like ignore `docs/example.[j|t]sx` files or any components ending in `Example` because I think this might be a problem for a lot of the devdocs components. 

Either way, it's definitely something for another PR.

tyxla commented on client/components/action-panel/figure.jsx:
```diff
@@ -8,10 +8,11 @@ import PropTypes from 'prop-types';
 	const figureClasses = classNames( {
 		'action-panel__figure': true,
 		'is-inline-body-text': inlineBodyText,
+		'is-aligned-left': 'left' === align,
```
How does this work with right-to-left languages? Seems like we'll end up with a `is-aligned-left` class with a `float:right`. Is that expected? Seems like it could be confusing.

Also I wonder if it makes more sense to add a `align-${ align }` classname instead to make this more robust.

pablinos replied:
Yep, that doesn't seem best semantically. Perhaps we can have an `align-first` and `align-last` or `align-before` and `align-after`?

tyxla replied:
No, don't get me wrong, I'm fine with it being `left` and `right`. "Before" and "after" might make it more confusing, at least in my head. What I meant we could use `align-${ align }` and add classes to style it properly. 

But also, I'm concerned about how things would behave if we hardcode alignment to be "left" and then we switch to an RTL language. Let's consider that too.

pablinos replied:
> But also, I'm concerned about how things would behave if we hardcode alignment to be "left" and then we switch to an RTL language. Let's consider that too.

Yes, that was why I was wondering if we shouldn't be using 'left' and 'right' in the class names. If something is on the left and we switch to an RTL language, would we want that element to be placed on the right? If so, we'd have a situation where `align-left` is styled to align something on the right, which seems pretty confusing and what it seemed like you were getting at with your original comment. 

I'll investigate what happens currently with RTL languages, but I think the figures will stay on the right, so it may make sense to leave them where they are for consistency.

tyxla commented on client/components/action-panel/style.scss:
```diff
@@ -103,8 +109,14 @@ a.action-panel__body-text-link {
 		}
 	}
+
+	.action-panel__body &::before {
```
Seems wrong that the footer is in the body here?

tyxla commented on client/components/action-panel/docs/example.jsx:
```diff
@@ -41,6 +41,30 @@ const ActionPanelExample = () => (
+						This is a description of the action. It gives a bit more detail and explains what we are
+						inviting the user to do.
+					</p>
+					<ActionPanelFooter>
```
Is the footer intentionally in the body? I'm not sure how much this makes sense from the semantics point of view.

tyxla commented on client/components/action-panel/README.md:
```diff
@@ -14,29 +14,53 @@ import ActionPanelFooter from 'components/action-panel/footer';
+						inviting the user to do.
+					</p>
+				</ActionPanelBody>
+				<ActionPanelFooter>
```
Seems odd to have the footer both in the body and outside of it in different examples. I think we should follow the same structure.

pablinos replied:
This is a tricky one to resolve. With the footer outside of the body and the figure on the left, the alignment of the buttons/links in the footer ends up under the image. This might be desired, but generally, where this pattern is being used the buttons are aligned with the body content.

We could change the HTML structure and/or not use floats, but we'd also need to audit where this component is being used and make sure that it's not affected. I'll do a first pass of that and see how much work it would be.

tyxla replied:
Yeah, I can see it gets complicated, but I'd like to think that if it's inside the body of element X, then it's not a footer of element X, it's something else. It could make more sense for it to be the footer of element Y, considering that element Y is nested inside X.

pablinos replied:
To keep things simple I've created another container component for use inside the body and changed the footer component to use it. I've called it `ActionPanelCta`, which might not be the best name, but in general, the footer is where the action links go, so it seemed appropriate.

tyxla reviewed (commented):
Things are looking good 👍 

Just left some comments and questions.

pablinos commented:
Thanks for taking a look at this @tyxla! I'll take a look at fixing the markup structure and semantics of the CSS classes.

pablinos committed:
Change alignment class name on the figure component

It now will have a specific `align-left` or `align-right` class name
applied.

pablinos committed:
Add new CTA component for use in the panel body

It was felt that the footer component shouldn't be used within the
`ActionPanel` body. Instead we have a generic CTA container component
(with shared CSS properties) and make the footer use that.

We also turn off the namespace class name lint check for the example.
